[Fix #520] Updating readme (#523)
serverlessworkflow/sdk-java: impl/README.md

    * Input
    * Output
* Expressions
    * Input
    * Output 
    * Export
    * Special keywords: runtime, workflow, task...
* Error definitions

## Setup

Serverless workflow reference implementation only requires setting up Java and Maven/Gradle. 

### JDK Version

Reference implementation requires [Java 17](https://openjdk.org/projects/jdk/17/) or newer versions. 

### Dependencies

One goal of the reference implementation is including the lowest number of dependencies possible. In that spirit, a modular approach has been used, letting the user decide which dependencies to include based on the nature of their workflow. 

In practical terms, this means establishing a separation between the core part and additional dependencies that should be explicitly included if your workflow is interacting with an external service using a particular technology supported by the specification (at this moment, just HTTP). The intention of this is to avoid adding dependencies that you do not need. For example, when gRPC call will be implemented, if we were adding the gRPC stack to the core dependencies, you won't be able to get rid of it even if none of your workflows use it. 

> #### Maven

Reviewer: Typo
```suggestion
### Maven
```

``` java 
try (WorkflowApplication appl = WorkflowApplication.builder().build()) {
      logger.info(
          "Workflow output is {}",
          appl.workflowDefinition(WorkflowReader.readWorkflowFromClasspath("get.yaml"))
              .instance(Map.of("petId", 10))
              .start()
              .join());
    }
```
You can find the complete java code [here](../examples/simpleGet/src/main/java/BlockingExample.java)

In order to execute the workflow without blocking the calling thread till the HTTP request is completed, you should write

``` java 
  try (WorkflowApplication appl = WorkflowApplication.builder().build()) {
      appl.workflowDefinition(WorkflowReader.readWorkflowFromClasspath("get.yaml"))
          .instance(Map.of("petId", 10))
          .start()
          .thenAccept(node -> logger.info("Workflow output is {}", node));
    }
```
When the HTTP request is done, both examples will print a similar output

```shell
Workflow output is {"id":10,"category":{"id":10,"name":"string"},"name":"doggie","photoUrls":["string"],"tags":[{"id":10,"name":"string"}],"status":"string"}

> 

Reviewer: Missing closing quotes  ```

You can find the complete java code [here](../examples/simpleGet/src/main/java/NotBlockingExample.java)

### Detailed version

To discuss runtime API we are going to use a couple of workflow:
- [listen.yaml](../examples/events/src/main/listen.yaml), which waits for an event reporting a temperature greater than 38
- [emit.yaml](../examples/events/src/main/emit.yaml), which emits events with a certain temperature, specified as workflow parameter.

> Here is a summary of what we are trying to do: We will start the `listen.yaml` workflow, which will be completed when it receives an event with the proper temperature. It won't block the main thread while waiting for it. Then, we will send an event with a lower temperature that will be ignored. Finally, we will send an event with a higher temperature to complete the waiting workflow. 

Reviewer: You can break down in several bullets:
- The listen.yaml workflow waits for an event (not-blocking).
- We send an event with a low temperature (ignored).
- We send an event with a high temperature (completes the workflow).

The first step is to create a [WorkflowApplication](core/src/main/java/io/serverlessworkflow/impl/WorkflowApplication.java) instance. An application is an abstraction that allows customization of different aspects of the workflow execution (for example, change the default `ExecutorService` for thread spawning)

> Since `WorkflowApplication` implements `Autocloseable`, we better use a try...finally block, ensuring any resource that the workflow might have used is freed when done. 

Reviewer: use try-with-resources term, as it's more accurate
```suggestion
Since `WorkflowApplication` implements `Autocloseable`, we better use a **try-with-resources** block, ensuring any resource that the workflow might have used is freed when done. 
```

```java
      WorkflowDefinition listenDefinition =
          appl.workflowDefinition(WorkflowReader.readWorkflowFromClasspath("listen.yaml"));
      WorkflowDefinition emitDefinition =
          appl.workflowDefinition(WorkflowReader.readWorkflowFromClasspath("emit.yaml")); 
```

A [WorkflowDefinition](core/src/main/java/io/serverlessworkflow/impl/WorkflowDefinition.java) object is immutable and, therefore, thread-safe. It is used to execute as many workflow instances as desired. 

To execute a workflow, we first create a [WorkflowInstance](core/src/main/java/io/serverlessworkflow/impl/WorkflowInstance.java) object (its initial status is PENDING) and then invoke the `start` method on it (its status is changed to RUNNING). The `start` method returns a [CompletableFuture](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html), which we use to indicate that a log message should be printed when the workflow is completed.

```java
  WorkflowInstance waitingInstance = listenDefinition.instance(Map.of());
      waitingInstance
          .start()
          .thenAccept(node -> logger.info("Waiting instance completed with result {}", node));
```

> As soon as the workflow execution reach the point where it waits for events to arrive, control is returned to the calling thread. Therefore, we can execute another workflow instance while the first one is waiting. 

Reviewer: The second phrase could start as "Since execution is non-blocking, ..." instead of "Therefore"

> We will send an event with a temperature that does not satisfy the criteria, so the listen instance will continue waiting. We use a regular Java ' Map ' to pass parameters to the workflow instance that sends the event. Note that since we want to wait till the event is published, we call `join` after `start`, telling the `CompletableFuture` to wait for workflow completion.

Reviewer: Fix the quotes around Map

# Serverless Workflow Specification - Java SDK- Reference Implementation

Welcome to Java SDK runtime reference implementation, a lightweight implementation of the Serverless Workflow specification which provides a simple, non blocking, reactive API for workflow execution. 

Although initially conceived mainly for testing purposes, it was designed to be easily expanded, so it can eventually become production ready. 

## Status. 

This reference implementation is currently capable of running workflows consisting of:

* Tasks
    * Switch 
    * Set
    * Do
    * Raise
    * Listen
    * Emit
    * Fork
    * For
    * Try
    * Wait
    * Call
        * HTTP
* Schema Validation
    * Input
    * Output
* Expressions
    * Input
    * Output 
    * Export
    * Special keywords: runtime, workflow, task...
* Error definitions

## Setup

Serverless workflow reference implementation only requires setting up Java and Maven/Gradle. 

### JDK Version

Reference implementation requires [Java 17](https://openjdk.org/projects/jdk/17/) or newer versions. 

### Dependencies

> One of the goals of the reference implemetantion is to maintain the number of dependencies as lower as possible. With that spirit, a modular approach has been followed, lettting the user decide, depending on their wokflows nature, which dependencies should be include. 

Reviewer 2: ```suggestion
One goal of the reference implementation is maintaining the lowest number of dependencies possible. In that spirit, a modular approach has been used, letting the user decide which dependencies to include based on the nature of their workflow. 
```

Author: Rephrased as proposed

> In practical terms, this means a separation between the core part and additional dependencies that should be explicitly included if your workflow is interacting with an external service that communicated using a particular technology supported by the specification (at this moment, just HTTP). The intention of this is to avoid adding dependencies that you do not really need (for example, when gRPC call will be implemented, if we were adding the gRPC stack to the core dependencies, you wont be able to get rid of it even if none of your workflows use it)

Reviewer 2: ```suggestion
In practical terms, this means a separation between the core part and additional dependencies that should be explicitly included if your workflow is interacting with an external service that communicated using a particular technology supported by the specification (at this moment, just HTTP). The intention of this is to avoid adding dependencies that you do not need (for example, when gRPC call will be implemented, if we were adding the gRPC stack to the core dependencies, you won't be able to get rid of it even if none of your workflows use it)
```

#### Maven

You always need to add this dependency to your pom.xml `dependencies` section:

```xml
<dependency>
      <groupId>io.serverlessworkflow</groupId>
      <artifactId>serverlessworkflow-impl-core</artifactId>
      <version>7.0.0-SNAPSHOT</version>
</dependency>
```

And only if your workflow is using HTTP calls, you must add:

```xml
<dependency>
      <groupId>io.serverlessworkflow</groupId>
      <artifactId>serverlessworkflow-impl-http</artifactId>
      <version>7.0.0-SNAPSHOT</version>
</dependency>
```

### Gradle projects:

You always need to add this dependency to your build.gradle `dependencies` section:

```text
implementation("io.serverlessworkflow:serverlessworkflow-impl-core:7.0.0-SNAPSHOT")
```

And only if your workflow is using HTTP calls, you must add:

```text
implementation("io.serverlessworkflow:serverlessworkflow-impl-http:7.0.0-SNAPSHOT")
```

## How to use

> Quick version is intended for impatient users that want to try something as soon as possible.

Reviewer 2: ```suggestion
The quick version is intended for impatient users who want to try something as soon as possible.
```

> Detailed version is more suitable for those users interested on a more thoughtful discussion of the API.

Reviewer 2: ```suggestion
The detailed version is more suitable for those users interested in a more thoughtful discussion of the API.
```

### Quick version

For a quick introduction, we are going to use a simple workflow [definition](../examples/simpleGet/src/main/resources/get.yaml) that performs a get call. 
We are going to show two ways of invoking the workflow: 
  - blocking the thread till the get request goes through
  - returning control to the caller, so the main thread continues while the get is executed

> In order to execute the workflow, blocking the thread till the http request is completed, you should write

Reviewer 2: ```suggestion
In order to execute the workflow, blocking the thread till the HTTP request is completed, you should write
```

``` java 
try (WorkflowApplication appl = WorkflowApplication.builder().build()) {
      logger.info(
          "Workflow output is {}",
          appl.workflowDefinition(WorkflowReader.readWorkflowFromClasspath("get.yaml"))
              .instance(Map.of("petId", 10))
              .start()
              .join());
    }
```
You can find the complete java code [here](../examples/simpleGet/src/main/java/BlockingExample.java)

> In order to execute the workflow, without blocking the calling thread till the http request is completed, you should write

Reviewer 2: ```suggestion
In order to execute the workflow without blocking the calling thread till the HTTP request is completed, you should write
```

```java
      WorkflowDefinition listenDefinition =
          appl.workflowDefinition(WorkflowReader.readWorkflowFromClasspath("listen.yaml"));
      WorkflowDefinition emitDefinition =
          appl.workflowDefinition(WorkflowReader.readWorkflowFromClasspath("emit.yaml")); 
```

A [WorkflowDefinition](core/src/main/java/io/serverlessworkflow/impl/WorkflowDefinition.java) object is immutable and therefore thread safe. It is used to execute as many workflow instances as desired. 

To execute a workflow, we first create a [WorkflowInstance](core/src/main/java/io/serverlessworkflow/impl/WorkflowInstance.java) object (the initial status is PENDING) and then invoke `start` method on it (the status is changed to RUNNING). `start` method returns a [CompletableFuture](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html), which we use to indicate that a log message should be printed when the workflow is completed.

```java
  WorkflowInstance waitingInstance = listenDefinition.instance(Map.of());
      waitingInstance
          .start()
          .thenAccept(node -> logger.info("Waiting instance completed with result {}", node));
```

> The next line will be executed as soon as the workflow execution starts waiting for events to arrive, moment at which control is returned to the calling thread. Therefore, we can execute another workflow instance while the first one is waiting. 

Author: In this case I wanter to say that it starts waiting, without comma

> The next line will be executed as soon as the workflow execution starts waiting for events to arrive, moment at which control is returned to the calling thread. Therefore, we can execute another workflow instance while the first one is waiting. 

Author: Rephrased
